File: platforms/ios/devkit/Sources/NativeblocksDevkit/feature/auth/data/AuthRepositoryImpl.swift
import Combine
import Foundation

internal class AuthRepositoryImpl: AuthRepository {
    private let cache: INativeCache
    private let TAG = "AuthRepository"
    init(cache: INativeCache) {
        self.cache = cache
    }

    func updateTokenAndEndpoint(qrData: String) -> Bool {
        do {
            let data = try JSONDecoder().decode(QrData.self, from: Data(qrData.utf8))
            cache.push(key: "ENDPOINT", value: data.endpoint)
            cache.push(key: "REALTIME_ENDPOINT", value: data.realtimeEndpoint)
            if let username = JWTUtils.decodedBody(data.token)?["email"] as? String {
                cache.push(key: "USERNAME", value: username)
            }
            cache.push(key: "UUID", value: UUID().uuidString)
            cache.push(key: "AUTH_TOKEN", value: data.token)
            return true
        } catch {
            DevKitLogger.error(TAG, "updateTokenAndEndpointQR", error)
            return false
        }
    }

    func updateTokenAndEndpoint(endpoint: String, token: String, realtimeEndpoint: String) -> Bool {
        cache.push(key: "ENDPOINT", value: endpoint)
        cache.push(key: "REALTIME_ENDPOINT", value: realtimeEndpoint)
        if let username = JWTUtils.decodedBody(token)?["email"] as? String {
            cache.push(key: "USERNAME", value: username)
        }
        cache.push(key: "UUID", value: UUID().uuidString)
        cache.push(key: "AUTH_TOKEN", value: token)
        return true
    }

    func authCheck() -> Bool {
        return !cache.pull(key: "AUTH_TOKEN", defaultValue: "").isEmpty
    }

    func logout() {
        cache.delete(key: "ENDPOINT")
        cache.delete(key: "AUTH_TOKEN")
        cache.delete(key: "REALTIME_ENDPOINT")
        cache.delete(key: "USERNAME")
        cache.delete(key: "UUID")
    }

    func token() -> AnyPublisher<String, Never> {
        return cache.onValueChange(key: "AUTH_TOKEN", defaultValue: "")
    }
}

File: platforms/android/devkit/devkit/src/main/java/io/nativeblocks/devkit/feature/auth/data/AuthRepositoryImpl.kt
package io.nativeblocks.devkit.feature.auth.data

import io.nativeblocks.devkit.feature.auth.domain.model.QrData
import io.nativeblocks.devkit.feature.auth.domain.repository.AuthRepository
import io.nativeblocks.devkit.lib.cache.INativeCache
import io.nativeblocks.devkit.util.JWT
import io.nativeblocks.devkit.util.DevKitLogger
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.json.Json
import java.util.UUID

internal class AuthRepositoryImpl(private val cache: INativeCache) : AuthRepository {

    override suspend fun updateTokenAndEndpoint(qrData: String): Boolean {
        return try {
            val json = Json { ignoreUnknownKeys = true }
            val data = json.decodeFromString<QrData>(qrData)
            cache.push("ENDPOINT", data.endpoint)
            cache.push("REALTIME_ENDPOINT", data.realtimeEndpoint)

            val decodedBody = JWT.decodedBody(data.token)
            val username = try {
                decodedBody?.getString("email")
            } catch (e: Exception) {
                ""
            }
            cache.push("USERNAME", username)

            cache.push("UUID", UUID.randomUUID().toString())
            cache.push("AUTH_TOKEN", data.token)
            true
        } catch (e: Exception) {
            DevKitLogger.e("AuthRepository", "update", e)
            false
        }
    }

    override suspend fun updateTokenAndEndpoint(
        endpoint: String,
        token: String,
        realtimeEndpoint: String
    ): Boolean {
        return try {
            cache.push("ENDPOINT", endpoint)
            cache.push("REALTIME_ENDPOINT", realtimeEndpoint)

            val decodedBody = JWT.decodedBody(token)
            val username = try {
                decodedBody?.getString("email")
            } catch (e: Exception) {
                ""
            }
            cache.push("USERNAME", username)

            cache.push("UUID", UUID.randomUUID().toString())
            cache.push("AUTH_TOKEN", token)
            true
        } catch (e: Exception) {
            DevKitLogger.e("AuthRepository", "update", e)
            false
        }
    }

    override suspend fun authCheck(): Boolean {
        return cache.pull("AUTH_TOKEN", "").isEmpty().not()
    }

    override suspend fun logout() {
        cache.delete("ENDPOINT")
        cache.delete("AUTH_TOKEN")
        cache.delete("REALTIME_ENDPOINT")
        cache.delete("USERNAME")
        cache.delete("UUID")
    }

    override fun token(): Flow<String> {
        return cache.onValueChange("AUTH_TOKEN", "")
    }
}